libbpf-tools/memleak.c

```diff
@@ -0,0 +1,707 @@
+		timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec, env.top);
+
+	ip = calloc(env.max_stack_depth, sizeof(*ip));
+	if (!ip) {
+		fprintf(stderr, "failed to alloccate memory of stack trace\n");
+		return;
+	}
+
+	stacks = calloc(MAX_HASH_ENTRY_NUM, sizeof(*stacks));
+	if (stacks == NULL) {
+		fprintf(stderr, "failed to allocate memory for stack info\n");
+		goto cleanup;
+	}
+	memset(stacks, 0, MAX_HASH_ENTRY_NUM * sizeof(*stacks));
+
+	cnt = 0;
+	ifd = bpf_map__fd(obj->maps.allocs);
+	sfd = bpf_map__fd(obj->maps.stack_traces);
+	while (!bpf_map_get_next_key(ifd, &addr, &next_addr)) {
+		err = bpf_map_lookup_elem(ifd, &next_addr, &info);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup info: %d\n", err);
+			goto cleanup;
+		}
+		addr = next_addr;
+		if (now_ns - env.min_age_ns < info.timestamp_ns) {
+			continue;
+		}
+
+		stack = NULL;
+		for (i=0; i < MAX_HASH_ENTRY_NUM; i++) {
+			if (stacks[i].stack_id == info.stack_id) {
+				stack = &stacks[i];
+				break;
+			}
+			if (stacks[i].stack_id == 0) {
+				stack = &stacks[i];
+				stack->stack_id = info.stack_id;
+				cnt++;
+				break;
+			}
+		}
+		if (stack == NULL) continue;
+		stack->size += info.size;
+		stack->nr++;
+		if (env.show_allocs) {
+			printf("\taddr = %lx size = %lld\n", addr, info.size);
+		}
+	}
+
+	quick_sort(stacks, 0, cnt);
+	if (cnt < max_prints) {
+		max_prints = cnt;
+	}
+
+	for (i=0; i < max_prints; i++) {
+		if (stacks[i].stack_id == 0) break;
+
+		if (bpf_map_lookup_elem(sfd, &stacks[i].stack_id, ip) != 0) {
+			fprintf(stderr, "    [Missed User Stack]\n");
+			continue;
+		}
+
+		if (env.kernel_trace) {
+			printf("\t%u bytes in %u allocations from stack\n", stacks[i].size, stacks[i].nr);
+			for (int j = 0; j < env.max_stack_depth && ip[j]; j++) {
+				ksym = ksyms__map_addr(ksyms, ip[j]);
+				if (ksym)
+					printf("\t\t%s (0x%lx)\n", ksym->name, ksym->addr);
+				else
+					printf("\t\t[unknown]\n");
+			}
+
+		} else {
+			syms = syms_cache__get_syms(syms_cache, env.pid);
+			if (!syms) {
+				fprintf(stderr, "failed to get syms\n");
+				continue;
+			}
+
+			printf("\t%u bytes in %u allocations from stack\n", stacks[i].size, stacks[i].nr);
+			for (int j = 0; j < env.max_stack_depth && ip[j]; j++) {
+				sym = syms__map_addr(syms, ip[j]);
+				if (sym)
+					printf("\t\t%s (0x%lx)\n", sym->name, sym->start);
+				else
+					printf("\t\t[unknown]\n");
+			}
+		}
+	}
+
+cleanup:
+	if (ip) free(ip);
+	if (stacks) free(stacks);
+}
+
+void sig_handler(int signo)
+{
+	if (env.child_pid == 0)
+		return;
+
+	if (signo == SIGINT || signo == SIGTERM) {
+		// kill all child processes forked by command
+		kill(0, SIGKILL);
+	}
+}
+
+int main(int argc, char **argv)
+{
+	char libc_path[PATH_MAX] = {};
+	struct memleak_bpf *skel;
+	struct syms_cache *syms_cache = NULL;
+	struct ksyms *ksyms = NULL;
+	int err, i;
+
+	/* Parse command line arguments */
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	if (env.command != NULL) {
```

Comment:
Move these to a new function ?

Reply:
Fix in commit [2beb289e](https://github.com/iovisor/bcc/pull/3828/commits/2beb289e98d651bddc2127299c5627b5d6149fce).